File: rate-limiter/python/tests/limiter_builder.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Optional

from rate_limiter import (
    DEFAULT_MAX_REQUESTS,
    DEFAULT_WINDOW_DURATION,
    Limiter,
    Rule,
)

from .fixed_clock import FixedClock


class LimiterBuilder:
    def __init__(self) -> None:
        self._max_requests: int = DEFAULT_MAX_REQUESTS
        self._window: timedelta = DEFAULT_WINDOW_DURATION
        self._start: datetime = datetime(2026, 1, 1, tzinfo=timezone.utc)
        self._clock: Optional[FixedClock] = None

    def with_max_requests(self, max_requests: int) -> "LimiterBuilder":
        self._max_requests = max_requests
        return self

    def with_window(self, window: timedelta) -> "LimiterBuilder":
        self._window = window
        return self

    def starting_at(self, start: datetime) -> "LimiterBuilder":
        self._start = start
        return self

    def with_clock(self, clock: FixedClock) -> "LimiterBuilder":
        self._clock = clock
        return self

    def build(self) -> tuple[Limiter, FixedClock]:
        clock = self._clock if self._clock is not None else FixedClock(self._start)
        limiter = Limiter(Rule(self._max_requests, self._window), clock)
        return limiter, clock

File: rate-limiter/typescript/tests/LimiterBuilder.ts
import {
  Limiter,
  DEFAULT_MAX_REQUESTS,
  DEFAULT_WINDOW_MS,
} from '../src/Limiter.js';
import { Rule } from '../src/Rule.js';
import { FixedClock } from './FixedClock.js';

export class LimiterBuilder {
  private maxRequests = DEFAULT_MAX_REQUESTS;
  private windowMs = DEFAULT_WINDOW_MS;
  private start = new Date(Date.UTC(2026, 0, 1));
  private clock: FixedClock | null = null;

  withMaxRequests(maxRequests: number): this {
    this.maxRequests = maxRequests;
    return this;
  }

  withWindowMs(windowMs: number): this {
    this.windowMs = windowMs;
    return this;
  }

  startingAt(start: Date): this {
    this.start = start;
    return this;
  }

  withClock(clock: FixedClock): this {
    this.clock = clock;
    return this;
  }

  build(): { limiter: Limiter; clock: FixedClock } {
    const clock = this.clock ?? new FixedClock(this.start);
    const limiter = new Limiter(new Rule(this.maxRequests, this.windowMs), clock);
    return { limiter, clock };
  }
}
